tests/conftest.py

```diff
@@ -0,0 +1,107 @@
+import pathlib
+import pytest
+import warnings
+from pytest_console_scripts import ScriptRunner
+from typing import Optional, Union
+
+
+class Project:
+    """
+    A Python project on which ``setup.py pyproject`` can be run. Tests should get access
+    to instances of this class by requesting the :py:func:`project()` fixture.
+
+    Instantiating ``Project`` creates the project under the given root directory, which
+    will typically be provided by pytest's ``tmp_path`` fixture. After instantiating
+    the ``Project`` object, create files underneath the root by calling methods like
+    :py:meth:`.setup_py()` and :py:meth:`.setup_cfg()`, or :py:meth:`.write()` for
+    custom filenames. Any unusual cases can be handled by directly operating on
+    :py:attr:`.root`. Finally, call :py:meth:`.run()` to run ``setup.py pyproject`` on
+    the project.
+
+    :param root:
+        The root directory in which to create the project. It should already exist.
+
+    :param script_runner:
+        The ``script_runner`` fixture from ``pytest-console-scripts``.
+    """
+
+    def __init__(self, root: pathlib.Path, script_runner: ScriptRunner) -> None:
+        self.root: pathlib.Path = root
+        self.script_runner: ScriptRunner = script_runner
+
+    def write(self, filename: Union[pathlib.Path, str], content: str) -> None:
+        """
+        Write a file with the given content and the given filename relative to
+        the project root. If the file already exists, it will be overwritten after
+        issuing a warning.
+
+        :param filename:
+            A filename or ``pathlib.Path`` representing the file to write. This should
+            be a relative path, which will be interpreted relative to the project root
+            directory. If the referenced file is not inside the project root, a warning
+            will be issued.
+
+        :param content:
+            Text content to write to the file.
+        """
+        file = (self.root / filename).resolve()
```

Comment:
Heh, I'm too addicted to `os.path`… I should really have a good close look at that `pathlib` library, that's much more intuitive than `os.path.join()` :-)